lib/index.js

```diff
@@ -47,78 +47,67 @@ module.exports = function loader (css, map) {
 
   validateOptions(require('./options.json'), options, 'PostCSS Loader')
 
-  const rc = {
-    path: path.dirname(file),
-    ctx: {
-      file: {
-        extname: path.extname(file),
-        dirname: path.dirname(file),
-        basename: path.basename(file)
-      },
-      options: {}
+  Promise.resolve().then(() => {
+    if (!options.config || (options.config && !options.config.path)) {
+      return parseOptions.call(this, options)
+    }
+
+    const rc = {
+      path: path.dirname(file),
+      ctx: {
+        file: {
+          extname: path.extname(file),
+          dirname: path.dirname(file),
+          basename: path.basename(file)
+        },
+        options: {}
+      }
     }
-  }
 
-  if (options.config) {
     if (options.config.path) {
       rc.path = path.resolve(options.config.path)
     }
 
     if (options.config.ctx) {
       rc.ctx.options = options.config.ctx
     }
-  }
-
-  const sourceMap = options.sourceMap
-
-  Promise.resolve().then(() => {
-    const length = Object.keys(options).length
-
-    // TODO
-    // Refactor
-    if (!options.config && !sourceMap && length) {
-      return parseOptions.call(this, options)
-    } else if (options.config && !sourceMap && length > 1) {
-      return parseOptions.call(this, options)
-    } else if (!options.config && sourceMap && length > 1) {
-      return parseOptions.call(this, options)
-    } else if (options.config && sourceMap && length > 2) {
-      return parseOptions.call(this, options)
-    }
 
     return postcssrc(rc.ctx, rc.path, { argv: false })
   }).then((config) => {
     if (!config) config = {}
 
     if (config.file) this.addDependency(config.file)
 
+    const sourceMap = options.sourceMap
+    delete options.sourceMap
+
     let plugins = config.plugins || []
-    let options = Object.assign({
+    let resolvedOptions = Object.assign({
       to: file,
       from: file,
       map: sourceMap
         ? sourceMap === 'inline'
           ? { inline: true, annotation: false }
           : { inline: false, annotation: false }
         : false
-    }, config.options)
+    }, options, config.options)
```

Comment:
> using options allow change map, it is bad, for some it is misleading behavior

? 🙃  You mean that currently one has to set `map` via `{ sourceMap: true || 'inline' }` (`webpack.config.js`) exclusively?

>  but to and from now also we can change through config.options

They shouldn't be changeable, since File I/O is handled by the webpack (Runner) 😛